### PR TITLE
net/local: Add Unix domain (local) sockets

### PR DESCRIPTION
**Pull request by junmin-kim**
Port UDS from Nuttx

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [84eb29e0e90d632fe6269076d8bfee09867fc71c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [84eb29e0e90d632fe6269076d8bfee09867fc71c] - Code Rule Check Result:
os/net/local/local.h:166: ERROR: [IDT_M_TAB] code indent should use tabs where possible
os/net/local/local.h:190: ERROR: [BRC_M_EUS] open brace '{' following struct go on the same line
os/net/local/local.h:199: ERROR: [BRC_M_EUS] open brace '{' following struct go on the same line
os/net/local/local.h:207: ERROR: [BRC_M_EUS] open brace '{' following struct go on the same line
os/net/local/local_sendpacket.c:83: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
os/net/local/local_sockif.c:113: ERROR: [BRC_M_SMT] that open brace { should be on the previous line
os/net/local/local_sockif.c:395: ERROR: [BRC_M_SMT] that open brace { should be on the previous line

**Comment by seinfra** [comment]
Target : [d39a3be3e16c04e3c26d16ac4e96e1f220441dc8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d39a3be3e16c04e3c26d16ac4e96e1f220441dc8] - Code Rule Check OK.

File: os/net/local/local_recvutils.c [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	ssize_t nread;
+	int ret;
+
+	DEBUGASSERT(buf && len);
+
+	remaining = *len;
+	while (remaining > 0) {
+		nread = file_read(filep, buf, remaining);
+		if (nread < 0) {
+			if (nread != -EINTR) {
```
Comment (zhouxinhe) [inline os/net/local/local_recvutils.c]: file_read() returns -1 on failure with errno set appropriately.
```suggestion
			if (errno != EINTR) {
```

File: os/net/local/local_sendpacket.c [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+static int local_fifo_write(FAR struct file *filep, FAR const uint8_t *buf,
+							size_t len)
+{
+	ssize_t nwritten;
+
+	while (len > 0) {
+		nwritten = file_write(filep, buf, len);
+		if (nwritten < 0) {
+			if (nwritten != -EINTR) {
```
Comment (zhouxinhe) [inline os/net/local/local_sendpacket.c]: file_write() returns -1 on failure with errno set appropriately.
```suggestion
			if (errno != EINTR) {
```

File: os/net/local/local_connect.c [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+	int sval;
+
+	/* Has server backlog been reached?
+	 * NOTE: The backlog will be zero if listen() has never been called by the
+	 * server.
+	 */
+
+	if (server->lc_state != LOCAL_STATE_LISTENING ||
+		server->u.server.lc_pending >= server->u.server.lc_backlog) {
+		net_unlock();
```
Comment (zhouxinhe) [inline os/net/local/local_connect.c]: net_unlock(...) require one parameter (even though it's not used).
```c
void net_unlock(net_lock_t flags)
```

File: os/net/local/local_recvutils.c (hunk 1 above)
Comment (junmin-kim) [reply os/net/local/local_recvutils.c]: I'll modify file_read

File: os/net/local/local_sendpacket.c (hunk 2 above)
Comment (junmin-kim) [reply os/net/local/local_sendpacket.c]: I'll also modify file_write

File: os/net/local/local_connect.c (hunk 3 above)
Comment (junmin-kim) [reply os/net/local/local_connect.c]: I replaced `net_unlock()` with `net_unlock(0)`

**Comment by seinfra** [comment]
Target : [c1a7434c69973979afea53ed4b86c603df25c61b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [c1a7434c69973979afea53ed4b86c603df25c61b] - Code Rule Check OK.

**Comment by juitem** [comment]
I think it would be better to replace the parameter '0' with meaningful letter or remove it. (later)

**Pull request merged**